src/openrct2/network/ServerList.cpp

```diff
@@ -80,3 +233,170 @@ bool server_list_write(const std::vector<server_entry>& entries)
         return false;
     }
 }
+
+std::future<std::vector<ServerListEntry>> ServerList::FetchLocalServerListAsync(const INetworkEndpoint& broadcastEndpoint)
+{
+    auto broadcastAddress = broadcastEndpoint.GetHostname();
+    return std::async(std::launch::async, [broadcastAddress] {
+        constexpr auto RECV_DELAY_MS = 10;
+        constexpr auto RECV_WAIT_MS = 2000;
+
+        std::string_view msg = NETWORK_LAN_BROADCAST_MSG;
+        auto udpSocket = CreateUdpSocket();
+
+        log_verbose("Broadcasting %zu bytes to the LAN (%s)", msg.size(), broadcastAddress.c_str());
+        auto len = udpSocket->SendData(broadcastAddress, NETWORK_LAN_BROADCAST_PORT, msg.data(), msg.size());
+        if (len != msg.size())
+        {
+            throw std::runtime_error("Unable to broadcast server query.");
+        }
+
+        std::vector<ServerListEntry> entries;
+        char buffer[1024]{};
+        size_t recievedLen{};
+        std::unique_ptr<INetworkEndpoint> endpoint;
+        for (int i = 0; i < (RECV_WAIT_MS / RECV_DELAY_MS); i++)
+        {
+            auto p = udpSocket->ReceiveData(buffer, sizeof(buffer) - 1, &recievedLen, &endpoint);
```

Comment:
Buffer is now always initialised to zeros prior to receive data leaving the remaining byte always 0 in case received data is not null terminated or too big. Invalid JSON is now handled by try...catch.